### PR TITLE
Update shard replication timestamp with max generated task

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
Update shard replication timestamp with max generated task

**Why?**
The namespace migration workflow use the task timestamp to calculate replication lag. We should use the timestamp from replication stack instead of real timestamp.

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/replication/ack_manager.go [hunk 1]
```diff
@@ -123,32 +124,42 @@ func (p *ackMgrImpl) NotifyNewTasks(
...
 	if p.maxTaskID == nil || *p.maxTaskID < maxTaskID {
 		p.maxTaskID = &maxTaskID
 	}
+	if p.maxTaskVisibilityTimestamp == nil || p.maxTaskVisibilityTimestamp.Before(maxVisibilityTimestamp) {
+		p.maxTaskVisibilityTimestamp = timestamp.TimePtr(maxVisibilityTimestamp)
+	}
 }
 
-func (p *ackMgrImpl) GetMaxTaskID() int64 {
+func (p *ackMgrImpl) GetMaxTaskInfo() (int64, time.Time) {
```
Comment (hehaifengcn) [inline service/history/replication/ack_manager.go:147]: nit: should we define a `struct` for return value?

File: service/history/replication/ack_manager.go [hunk 2]
```diff
@@ -123,32 +124,42 @@ func (p *ackMgrImpl) NotifyNewTasks(
...
+	// MaxTaskID and MaxTaskVisibilityTimestamp should be set in the same place.
+	if p.maxTaskID == nil || p.maxTaskVisibilityTimestamp == nil {
 		// maxTaskID is nil before any replication task is written which happens right after shard reload. In that case,
 		// use ImmediateTaskMaxReadLevel which is the max task id of any immediate task queues.
 		// ImmediateTaskMaxReadLevel will be the lower bound of new range_id if shard reload. Remote cluster will quickly (in
 		// a few seconds) ack to the latest ImmediateTaskMaxReadLevel if there is no replication tasks at all.
-		return p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
+		maxTaskID := p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
+		maxVisibilityStampe := p.shard.GetTimeSource().Now()
+		return maxTaskID, maxVisibilityStampe
```
Comment (hehaifengcn) [inline service/history/replication/ack_manager.go:159]: Not sure if it is possible that `p.maxTaskID` is nil while `p.maxTaskVisibilityTimestamp` (and inverse case too) is not but current code will skip non nil value. Shall we instead

```
maxTaskID := p.maxTaskID
if maxTaskID == nil {
    maxTaskID =  p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
}
```
same for `maxVisibilityStamp`

File: service/history/replication/ack_manager.go [hunk 3]
```diff
@@ -123,32 +124,42 @@ func (p *ackMgrImpl) NotifyNewTasks(
...
-	if p.maxTaskID == nil {
+	// MaxTaskID and MaxTaskVisibilityTimestamp should be set in the same place.
+	if p.maxTaskID == nil || p.maxTaskVisibilityTimestamp == nil {
 		// maxTaskID is nil before any replication task is written which happens right after shard reload. In that case,
 		// use ImmediateTaskMaxReadLevel which is the max task id of any immediate task queues.
 		// ImmediateTaskMaxReadLevel will be the lower bound of new range_id if shard reload. Remote cluster will quickly (in
 		// a few seconds) ack to the latest ImmediateTaskMaxReadLevel if there is no replication tasks at all.
-		return p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
+		maxTaskID := p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
+		maxVisibilityStampe := p.shard.GetTimeSource().Now()
```
Comment (hehaifengcn) [inline service/history/replication/ack_manager.go:158]: typo: `maxVisibilityTimestamp`

File: service/worker/migration/activities.go [hunk 4]
```diff
@@ -121,7 +121,7 @@ func (a *activities) checkReplicationOnce(ctx context.Context, waitRequest waitR
 			// Caught up to the last checked IDs, and within allowed lagging range
 			if clusterInfo.AckedTaskId >= waitRequest.WaitForTaskIds[shard.ShardId] &&
 				(shard.MaxReplicationTaskId-clusterInfo.AckedTaskId <= waitRequest.AllowedLaggingTasks ||
-					shard.ShardLocalTime.Sub(*clusterInfo.AckedTaskVisibilityTime) <= waitRequest.AllowedLagging) {
+					shard.MaxReplicationTaskVisibilityTime.Sub(*clusterInfo.AckedTaskVisibilityTime) <= waitRequest.AllowedLagging) {
```
Comment (hehaifengcn) [inline service/worker/migration/activities.go:124]: do we have any integration tests which can validate the result here?

**Review by hehaifengcn — APPROVED** [review]

File: service/history/replication/ack_manager.go (hunk 3 above)
Comment (yux0) [reply service/history/replication/ack_manager.go:158]: Fixed

File: service/history/replication/ack_manager.go (hunk 2 above)
Comment (yux0) [reply service/history/replication/ack_manager.go:159]: Updated

File: service/worker/migration/activities.go (hunk 4 above)
Comment (yux0) [reply service/worker/migration/activities.go:124]: Added an unit test in ack manager. The e2e is covered with int test.

File: service/history/replication/ack_manager.go (hunk 1 above)
Comment (yux0) [reply service/history/replication/ack_manager.go:147]: don't think it is necessary here.

**Review by yycptt — APPROVED** [review]

**Pull request merged**